warehouse/macaroons/caveats/__init__.py

```diff
@@ -104,7 +104,7 @@ class RequestUser(Caveat):
     user_id: StrictStr
 
     def verify(self, request: Request, context: Any, permission: str) -> Result:
-        if not isinstance(request.identity, User):
+        if not isinstance(request.identity, UserTokenContext):
```

Comment:
NB: This will address https://github.com/pypi/warehouse/pull/15588#issuecomment-1996036127